### PR TITLE
make ADIOS dependency on external MXML optional

### PR DESCRIPTION
**Pull request by olesenm**
- Follow the default ADIOS setup and use the version included in the
  source tree (as per ZFP) unless otherwise selected.

- Can also circumvent some spurious build issues experienced with mxml

- python patch not applicable to adios 1.9.0

### CODE REVIEW

File: var/spack/repos/builtin/packages/adios/package.py [hunk 1]
```diff
@@ -50,11 +50,12 @@ class Adios(AutotoolsPackage):
 
     variant('mpi', default=True, description='Enable MPI support')
     variant('infiniband', default=False, description='Enable infiniband support')
+    variant('mxml', default=False, description='Use external MXML library')
```
Comment (davydden) [inline var/spack/repos/builtin/packages/adios/package.py]: i would keep this `True`

File: var/spack/repos/builtin/packages/adios/package.py [hunk 2]
```diff
@@ -50,11 +50,12 @@ class Adios(AutotoolsPackage):
...
 
     # transforms
     variant('zlib', default=True, description='Enable zlib transform support')
     variant('szip', default=False, description='Enable szip transform support')
-    variant('zfp', default=False, description='Enable ZFP transform support')
+    variant('zfp', default=False, description='Use external ZFP library')
```
Comment (davydden) [inline var/spack/repos/builtin/packages/adios/package.py]: same here, `True`

**Comment by davydden** [comment]
> circumvent some spurious build issues experienced with mxml

i think this should be addressed with `mxml` instead of make packages which use it build internal versions.

File: var/spack/repos/builtin/packages/adios/package.py (hunk 2 above)
Comment (olesenm) [inline var/spack/repos/builtin/packages/adios/package.py]: Sure, but without +zfp ADIOS will take its internal version anyhow. So even with default=True, the description is still "Use external ZFP" and not "enable/disable ZFP".
Similarly for MXML...

So then it is really just a ma

File: var/spack/repos/builtin/packages/adios/package.py (hunk 1 above)
Comment (olesenm) [inline var/spack/repos/builtin/packages/adios/package.py]: I wouldn't actually care, if the mxml configure and build were not broken. It is extremely tiresome to fix N different dependent packages just to get the 1 package actually built!

**Comment by olesenm** [comment]
OK, so change MXML back to True (and hope that someone else will take a look at the mxml build) and we can commit this?
The new description for +zfp is still accurate. Having adios~zfp will still have zfp support, just not via an external library.

File: var/spack/repos/builtin/packages/adios/package.py (hunk 2 above)
Comment (davydden) [inline var/spack/repos/builtin/packages/adios/package.py]: > the description is still "Use external ZFP"

sure

File: var/spack/repos/builtin/packages/adios/package.py (hunk 1 above)
Comment (davydden) [inline var/spack/repos/builtin/packages/adios/package.py]: > It is extremely tiresome to fix N different dependent packages just to get the 1 package actually built!

i know the pain, and I came across similar issues while building `dealii` which eventually depends on another 20+ packages. My take on this problem is that this is the price to pay when working with Spack. The Spack community do its best to maintain a working collection of packages across various platforms/compilers/variants/etc. And if something is broken for specific case, then it's ought to be fixed anyhow.

**Comment by davydden** [comment]
> OK, so change MXML back to True

you probably forgot to push, I don't see any changes to this PR.

**Comment by davydden** [comment]
@olesenm 

> I wouldn't actually care, if the mxml configure and build were not broken.

btw, mind creating an issue with the details of what is broken on what system with what compiler and what errors etc?

**Comment by olesenm** [comment]
Added as issue #3700 - during which I see that the ADIOS package wasn't quite correct in the first place.
Since mxml builds a static anyhow, the external mxml dependency should really just be a _build_ dependency for ADIOS.

This opens up the question about how to pass build-only parameters? That is to say, if we use an external mxml static library to build adios, the resulting .so doesn't depend on it. So adios+mxml isn't a real adios variant, but more of a build specification (eg, like `spack install -j 4 adios`).

**Comment by olesenm** [comment]
After making the changes, it doesn't look like adios+shared builds a shared library anyhow, so I think it is quite immaterial if we build from static sources included as part of ADIOS or use a static library provided by MXML (when its pull-request gets merged).

**Comment by adamjstewart** [comment]
@olesenm Is the optional `mxml` variant still necessary now that #3706 is merged? Even if it is, we can still merge the smaller changes like the patch version range.

**Comment by olesenm** [comment]
Hi @adamjstewart - with the merge, the mxml variant is not strictly needed, but still seems reasonable (to me) to distinguish between external mxml (static!) and internal mxml linkage ... just like the zfp variant is actually internal vs external sources. But if the mxml variant bugs to many people, I'm OK with that too.

**Comment by olesenm** [comment]
This is quite obviously going nowhere, I'll close this.

**Pull request closed without merging**